talk/expert/coroutines.tex

```diff
@@ -0,0 +1,413 @@
+\subsection{Coroutines}
+
+\begin{frame}[fragile]
+  \frametitlecpp[20]{Why do we need coroutines ?}
+  \begin{block}{The generator use case (one of several)}
+    In python one can write
+    {\scriptsize
+      \begin{pythoncode*}{gobble=2}
+        for i in range(0, 10):
+          print(i)
+      \end{pythoncode*}
+    }
+    What about it in \cpp ?
+    {\scriptsize
+      \begin{cppcode*}{gobble=2, firstnumber=3}
+        someType range(int first, int last) { ... }
+        for (auto i : range(0, 10)) {
+          std::cout << i << "\n";
+        }
+      \end{cppcode*}
+    }
+    How can we implement range ?
+  \end{block}
```

Comment:
This example is not ideal, because `range(0, 10)` could just be implemented as a class with iterators. It is effectively `std::view::iota(0, 10)`.

The big selling point of generator coroutines is that the collection is built lazy and the full range of elements needs never to be stored. Plus, the syntax for the user is easy (because we could always implement it as a range as well).

A more interesting example is this one here: https://stackoverflow.com/a/71193067. A coroutine that generates fibonacci numbers. Here is a simplified version:

```c++
std::generator<int> fib() {
  co_yield 0;
  co_yield 1;
  int a=0;
  int b=1;
  while(true) {
    const auto next = a + b;
    co_yield next;
    a = b;
    b = next;
  }
}
```
Try to write this using a normal range class and iterators! :)

And then use like this:
```c++
std::print("{}", fib() | std::ranges::take(17));
```
(`std::print` is new in C++23 and will be able to print ranges)

Reply:
I see from the following slides, how you are using this example to build up coroutines. So maybe leave it here then. You may want to include my `fib` example though, as it shows more control flow inside a coroutine.

Reply:
The example is to simple indeed, but it helps to start simple. Now I was already wondering whether to add the fibonacci example as a kind of recap at the end. Then I will, as a non trivial example where corutines do bring something. Btw, it's there in cppreference too : https://en.cppreference.com/w/cpp/language/coroutines towards the end

Note : I'm away next week, so will conclude on that and ranges only in 10 days

Reply:
I've now added the Fibonacci example, revisited the rest and extended it with some slides on awaiters and awaitable. Should be close to final version